File: workbook/ch08/data/redblacktree/main.py
class Node:
    def __init__(self, key, value=None, color=1):
        self.key = key
        self.value = value
        self.parent = None
        self.left = None
        self.right = None
        self.color = color  # 1 for RED, 0 for BLACK

class RedBlackTree:
    def __init__(self):
        self.NIL = Node(None, None, 0)  # sentinel node (nil), colored BLACK
        self.root = self.NIL

    def search(self, key):
        return self._search_helper(self.root, key)

    def _search_helper(self, node, key):
        if node == self.NIL:
            return None
        
        if key == node.key:
            return node.value
        elif key < node.key:
            return self._search_helper(node.left, key)
        else:
            return self._search_helper(node.right, key)

    def insert(self, key, value=None):
        # Create new node
        new_node = Node(key, value, 1)  # RED node
        new_node.left = self.NIL
        new_node.right = self.NIL
        
        # Find position to insert
        y = None
        x = self.root
        
        while x != self.NIL:
            y = x
            if new_node.key < x.key:
                x = x.left
            elif new_node.key > x.key:
                x = x.right
            else:
                # Key already exists, update value
                x.value = value
                return
        
        # Set parent of new node
        new_node.parent = y
        
        # Insert new node
        if y is None:
            self.root = new_node
        elif new_node.key < y.key:
            y.left = new_node
        else:
            y.right = new_node
            
        # Fix Red-Black Tree properties
        self._fix_insert(new_node)

    def _fix_insert(self, k):
        while k != self.root and k.parent.color == 1:  # RED
            if k.parent == k.parent.parent.right:
                u = k.parent.parent.left  # uncle
                if u.color == 1:  # RED
                    u.color = 0  # BLACK
                    k.parent.color = 0  # BLACK
                    k.parent.parent.color = 1  # RED
                    k = k.parent.parent
                else:
                    if k == k.parent.left:
                        k = k.parent
                        self._right_rotate(k)
                    k.parent.color = 0  # BLACK
                    k.parent.parent.color = 1  # RED
                    self._left_rotate(k.parent.parent)
            else:
                u = k.parent.parent.right  # uncle
                if u.color == 1:  # RED
                    u.color = 0  # BLACK
                    k.parent.color = 0  # BLACK
                    k.parent.parent.color = 1  # RED
                    k = k.parent.parent
                else:
                    if k == k.parent.right:
                        k = k.parent
                        self._left_rotate(k)
                    k.parent.color = 0  # BLACK
                    k.parent.parent.color = 1  # RED
                    self._right_rotate(k.parent.parent)
            
            if k == self.root:
                break
                
        self.root.color = 0  # BLACK

    def delete(self, key):
        self._delete_node(self.root, key)

    def _delete_node(self, node, key):
        z = self.NIL
        while node != self.NIL:
            if node.key == key:
                z = node
                break
            elif node.key > key:
                node = node.left
            else:
                node = node.right
                
        if z == self.NIL:
            return False  # Key not found
            
        y = z
        y_original_color = y.color
        
        if z.left == self.NIL:
            x = z.right
            self._transplant(z, z.right)
        elif z.right == self.NIL:
            x = z.left
            self._transplant(z, z.left)
        else:
            y = self._minimum(z.right)
            y_original_color = y.color
            x = y.right
            
            if y.parent == z:
                x.parent = y
            else:
                self._transplant(y, y.right)
                y.right = z.right
                y.right.parent = y
                
            self._transplant(z, y)
            y.left = z.left
            y.left.parent = y
            y.color = z.color
            
        if y_original_color == 0:  # BLACK
            self._fix_delete(x)
            
        return True

    def _fix_delete(self, x):
        while x != self.root and x.color == 0:  # BLACK
            if x == x.parent.left:
                s = x.parent.right
                if s.color == 1:  # RED
                    s.color = 0  # BLACK
                    x.parent.color = 1  # RED
                    self._left_rotate(x.parent)
                    s = x.parent.right
                    
                if s.left.color == 0 and s.right.color == 0:  # Both BLACK
                    s.color = 1  # RED
                    x = x.parent
                else:
                    if s.right.color == 0:  # BLACK
                        s.left.color = 0  # BLACK
                        s.color = 1  # RED
                        self._right_rotate(s)
                        s = x.parent.right
                        
                    s.color = x.parent.color
                    x.parent.color = 0  # BLACK
                    s.right.color = 0  # BLACK
                    self._left_rotate(x.parent)
                    x = self.root
            else:
                s = x.parent.left
                if s.color == 1:  # RED
                    s.color = 0  # BLACK
                    x.parent.color = 1  # RED
                    self._right_rotate(x.parent)
                    s = x.parent.left
                    
                if s.right.color == 0 and s.left.color == 0:  # Both BLACK
                    s.color = 1  # RED
                    x = x.parent
                else:
                    if s.left.color == 0:  # BLACK
                        s.right.color = 0  # BLACK
                        s.color = 1  # RED
                        self._left_rotate(s)
                        s = x.parent.left
                        
                    s.color = x.parent.color
                    x.parent.color = 0  # BLACK
                    s.left.color = 0  # BLACK
                    self._right_rotate(x.parent)
                    x = self.root
                    
        x.color = 0  # BLACK

    def _transplant(self, u, v):
        if u.parent is None:
            self.root = v
        elif u == u.parent.left:
            u.parent.left = v
        else:
            u.parent.right = v
        v.parent = u.parent

    def _minimum(self, node):
        while node.left != self.NIL:
            node = node.left
        return node

    def _left_rotate(self, x):
        y = x.right
        x.right = y.left
        
        if y.left != self.NIL:
            y.left.parent = x
            
        y.parent = x.parent
        
        if x.parent is None:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
            
        y.left = x
        x.parent = y

    def _right_rotate(self, x):
        y = x.left
        x.left = y.right
        
        if y.right != self.NIL:
            y.right.parent = x
            
        y.parent = x.parent
        
        if x.parent is None:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
            
        y.right = x
        x.parent = y

    def inorder(self):
        result = []
        self._inorder_helper(self.root, result)
        return result

    def _inorder_helper(self, node, result):
        if node != self.NIL:
            self._inorder_helper(node.left, result)
            result.append((node.key, node.value, "RED" if node.color == 1 else "BLACK"))
            self._inorder_helper(node.right, result)

    def is_empty(self):
        return self.root == self.NIL

def test_red_black_tree():
    # Test Case 0: Insert nodes
    print("Test Case 0: Insert nodes")
    tree = RedBlackTree()
    tree.insert("Athena", "Wisdom")
    tree.insert("Zeus", "Thunder")
    tree.insert("Hades", "Underworld")
    tree.insert("Poseidon", "Sea")
    tree.insert("Hermes", "Messenger")
    tree.insert("Demeter", "Harvest")
    tree.insert("Dionysus", "Wine")
    tree.insert("Ares", "War")
    tree.insert("Artemis", "Hunt")
    tree.insert("Hephaestus", "Forge")
    print("Inorder Walk:")
    print(tree.inorder())
    print()

    # Create a new tree for the rest of the test cases
    tree = RedBlackTree()

    # Test Case 1: Empty Tree
    print("Test Case 1: Empty Tree")
    print(f"Search for 'Athena': {'Not Found' if tree.search('Athena') is None else 'Found'}")
    tree.delete("Athena")  # Should do nothing
    print()

    # Test Case 2: Single Node Tree
    print("Test Case 2: Single Node Tree")
    tree.insert("Athena", "Wisdom")
    print("Inorder Walk:")
    print(tree.inorder())
    print(f"Search for 'Athena': {'Not Found' if tree.search('Athena') is None else 'Found'}")
    tree.delete("Athena")  # Delete the root
    print("Inorder Walk after deletion:")
    print(tree.inorder())
    print()

    # Test Case 3: Insertion of Duplicate Keys
    print("Test Case 3: Insertion of Duplicate Keys")
    tree.insert("Athena", "Wisdom")
    tree.insert("Athena", "New Value")  # Update value
    print("Inorder Walk:")
    print(tree.inorder())
    print()

    # Test Case 4: Deletion of Non-Existent Keys
    print("Test Case 4: Deletion of Non-Existent Keys")
    tree.delete("NonExistentKey")  # Should do nothing
    print("Inorder Walk:")
    print(tree.inorder())
    print()

    # Test Case 5: Deletion of Root Node
    print("Test Case 5: Deletion of Root Node")
    tree.insert("Zeus", "Thunder")
    tree.insert("Hades", "Underworld")
    print("Inorder Walk before deletion:")
    print(tree.inorder())
    print()
    tree.delete("Hades")
    print("Inorder Walk after deletion:")
    print(tree.inorder())
    print()

    # Test Case 6: Deletion of Leaf Nodes
    print("Test Case 6: Deletion of Leaf Nodes")
    tree.insert("Poseidon", "Sea")
    print("Inorder Walk before deletion:")
    print(tree.inorder())
    print()
    tree.delete("Athena")  # Delete leaf node
    print("Inorder Walk after deletion:")
    print(tree.inorder())
    print()

    # Test Case 7: Deletion of Nodes with One Child
    print("Test Case 7: Deletion of Nodes with One Child")
    tree.insert("Hermes", "Messenger")
    tree.insert("Demeter", "Harvest")
    print("Inorder Walk before deletion:")
    print(tree.inorder())
    print()
    tree.delete("Hermes")  # Delete node with one child
    print("Inorder Walk after deletion:")
    print(tree.inorder())
    print()

    # Test Case 8: Deletion of Nodes with Two Children
    print("Test Case 8: Deletion of Nodes with Two Children")
    tree.insert("Dionysus", "Wine")
    tree.insert("Ares", "War")
    print("Inorder Walk before deletion:")
    print(tree.inorder())
    print()
    tree.delete("Demeter")  # Delete node with two children
    print("Inorder Walk after deletion:")
    print(tree.inorder())
    print()

    # Test Case 9: Large Tree Stress Test
    print("Test Case 9: Large Tree Stress Test")
    for i in range(100):
        key = f"Key{i}"
        value = f"Value{i}"
        tree.insert(key, value)
    
    print("Inorder Walk after insertions (showing only first 10 items):")
    full_inorder = tree.inorder()
    print(full_inorder[:10])
    print("...")
    
    for i in range(100):
        key = f"Key{i}"
        tree.delete(key)
    
    print("Inorder Walk after deletions:")
    print(tree.inorder())
    print()

if __name__ == "__main__":
    test_red_black_tree()

File: workbook/ch08/data/redblacktree/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// colors for Red-Black Tree
#define RED 0
#define BLACK 1

// Red-Black Tree Node
typedef struct RedBlackTreeNode {
    char *key;
    char *value;
    int color;
    struct RedBlackTreeNode *left, *right, *parent;
} RedBlackTreeNode;

// Red-Black Tree Structure
typedef struct {
    RedBlackTreeNode *root;
    RedBlackTreeNode *nil; // sentinel node for leaves
} RedBlackTree;


RedBlackTreeNode *createNode(char *key, char *value) {
    RedBlackTreeNode *node = (RedBlackTreeNode *)malloc(sizeof(RedBlackTreeNode));
    node->key = strdup(key);
    node->value = strdup(value);
    node->color = RED;
    node->left = node->right = node->parent = NULL;
    return node;
}

RedBlackTree *createRedBlackTree() {
    RedBlackTree *tree = (RedBlackTree *)malloc(sizeof(RedBlackTree));
    tree->nil = createNode("Nil", "Nil");
    tree->nil->color = BLACK;
    tree->root = tree->nil;
    return tree;
}

int comp(char *key1, char *key2) {
    return strcmp(key1, key2);
}

void inorderWalk(RedBlackTreeNode *x, RedBlackTreeNode *nil) {
    if (x != nil) {
        inorderWalk(x->left, nil);
        printf("Key: %s, Value: %s\n", x->key, x->value);
        inorderWalk(x->right, nil);
    }
}

void rotateLeft(RedBlackTree *tree, RedBlackTreeNode *x) {
    RedBlackTreeNode *y = x->right;
    x->right = y->left;
    if (y->left != tree->nil) {
        y->left->parent = x;
    }
    y->parent = x->parent;
    if (x->parent == tree->nil) {
        tree->root = y;
    } else if (x == x->parent->left) {
        x->parent->left = y;
    } else {
        x->parent->right = y;
    }
    y->left = x;
    x->parent = y;
}

void rotateRight(RedBlackTree *tree, RedBlackTreeNode *y) {
    RedBlackTreeNode *x = y->left;
    y->left = x->right;
    if (x->right != tree->nil) {
        x->right->parent = y;
    }
    x->parent = y->parent;
    if (y->parent == tree->nil) {
        tree->root = x;
    } else if (y == y->parent->left) {
        y->parent->left = x;
    } else {
        y->parent->right = x;
    }
    x->right = y;
    y->parent = x;
}

void insertFixup(RedBlackTree *tree, RedBlackTreeNode *z) {
    while (z->parent->color == RED) {
        if (z->parent == z->parent->parent->left) {
            RedBlackTreeNode *y = z->parent->parent->right;
            if (y->color == RED) {
                z->parent->color = BLACK;
                y->color = BLACK;
                z->parent->parent->color = RED;
                z = z->parent->parent;
            } else {
                if (z == z->parent->right) {
                    z = z->parent;
                    rotateLeft(tree, z);
                }
                z->parent->color = BLACK;
                z->parent->parent->color = RED;
                rotateRight(tree, z->parent->parent);
            }
        } else {
            RedBlackTreeNode *y = z->parent->parent->left;
            if (y->color == RED) {
                z->parent->color = BLACK;
                y->color = BLACK;
                z->parent->parent->color = RED;
                z = z->parent->parent;
            } else {
                if (z == z->parent->left) {
                    z = z->parent;
                    rotateRight(tree, z);
                }
                z->parent->color = BLACK;
                z->parent->parent->color = RED;
                rotateLeft(tree, z->parent->parent);
            }
        }
    }
    tree->root->color = BLACK;
}

void insert(RedBlackTree *tree, RedBlackTreeNode *z) {
    RedBlackTreeNode *y = tree->nil;
    RedBlackTreeNode *x = tree->root;
    while (x != tree->nil) {
        y = x;
        if (comp(z->key, x->key) < 0) {
            x = x->left;
        } else if (comp(z->key, x->key) > 0) {
            x = x->right;
        } else {
            free(x->value);
            x->value = strdup(z->value);
            free(z->key);
            free(z->value);
            free(z);
            return;
        }
    }
    z->parent = y;
    if (y == tree->nil) {
        tree->root = z;
    } else if (comp(z->key, y->key) < 0) {
        y->left = z;
    } else {
        y->right = z;
    }
    z->left = tree->nil;
    z->right = tree->nil;
    z->color = RED;
    insertFixup(tree, z);
}

// find minimum node in subtree
RedBlackTreeNode *minimum(RedBlackTree *tree, RedBlackTreeNode *x) {
    while (x->left != tree->nil) {
        x = x->left;
    }
    return x;
}

// transplant a subtree
void transplant(RedBlackTree *tree, RedBlackTreeNode *u, RedBlackTreeNode *v) {
    if (u->parent == tree->nil) {
        tree->root = v;
    } else if (u == u->parent->left) {
        u->parent->left = v;
    } else {
        u->parent->right = v;
    }
    v->parent = u->parent;
}

// fix the tree after deletion
void deleteFixup(RedBlackTree *tree, RedBlackTreeNode *x) {
    while (x != tree->root && x->color == BLACK) {
        if (x == x->parent->left) {
            RedBlackTreeNode *w = x->parent->right;
            if (w->color == RED) {
                w->color = BLACK;
                x->parent->color = RED;
                rotateLeft(tree, x->parent);
                w = x->parent->right;
            }
            if (w->left->color == BLACK && w->right->color == BLACK) {
                w->color = RED;
                x = x->parent;
            } else {
                if (w->right->color == BLACK) {
                    w->left->color = BLACK;
                    w->color = RED;
                    rotateRight(tree, w);
                    w = x->parent->right;
                }
                w->color = x->parent->color;
                x->parent->color = BLACK;
                w->right->color = BLACK;
                rotateLeft(tree, x->parent);
                x = tree->root;
            }
        } else {
            RedBlackTreeNode *w = x->parent->left;
            if (w->color == RED) {
                w->color = BLACK;
                x->parent->color = RED;
                rotateRight(tree, x->parent);
                w = x->parent->left;
            }
            if (w->right->color == BLACK && w->left->color == BLACK) {
                w->color = RED;
                x = x->parent;
            } else {
                if (w->left->color == BLACK) {
                    w->right->color = BLACK;
                    w->color = RED;
                    rotateLeft(tree, w);
                    w = x->parent->left;
                }
                w->color = x->parent->color;
                x->parent->color = BLACK;
                w->left->color = BLACK;
                rotateRight(tree, x->parent);
                x = tree->root;
            }
        }
    }
    x->color = BLACK;
}

// delete a node from the tree
void delete(RedBlackTree *tree, RedBlackTreeNode *z) {
    RedBlackTreeNode *y = z;
    RedBlackTreeNode *x;
    int yOriginalColor = y->color;

    if (z->left == tree->nil) {
        x = z->right;
        transplant(tree, z, z->right);
    } else if (z->right == tree->nil) {
        x = z->left;
        transplant(tree, z, z->left);
    } else {
        y = minimum(tree, z->right);
        yOriginalColor = y->color;
        x = y->right;
        if (y->parent == z) {
            x->parent = y;
        } else {
            transplant(tree, y, y->right);
            y->right = z->right;
            y->right->parent = y;
        }
        transplant(tree, z, y);
        y->left = z->left;
        y->left->parent = y;
        y->color = z->color;
    }

    if (yOriginalColor == BLACK) {
        deleteFixup(tree, x);
    }

    free(z->key);
    free(z->value);
    free(z);
}

// search for a node by key
RedBlackTreeNode *search(RedBlackTree *tree, RedBlackTreeNode *x, char *key) {
    if (x == tree->nil || strcmp(key, x->key) == 0) {
        return x;
    }
    if (strcmp(key, x->key) < 0) {
        return search(tree, x->left, key);
    } else {
        return search(tree, x->right, key);
    }
}

// delete a node by key
void deleteByKey(RedBlackTree *tree, char *key) {
    RedBlackTreeNode *z = search(tree, tree->root, key);
    if (z != tree->nil) {
        delete(tree, z);
    } else {
        printf("Key not found: %s\n", key);
    }
}

void freeNode(RedBlackTree *tree, RedBlackTreeNode *node) {
    if (node == tree->nil) {
        return;
    }
    freeNode(tree, node->left);
    freeNode(tree, node->right);
    free(node->key);
    free(node->value);
    free(node);
}

void freeRedBlackTree(RedBlackTree *tree) {
    if (tree == NULL) {
        return;
    }
    freeNode(tree, tree->root);
    free(tree->nil->key);
    free(tree->nil->value);
    free(tree->nil);
    free(tree);
}


void printTreeHelper(RedBlackTreeNode *root, RedBlackTreeNode *nil, int space) {
    if (root == nil) {
        return;
    }

    // Increase distance between levels
    space += 10;

    // Process right child first (above the parent)
    printTreeHelper(root->right, nil, space);

    // Print current node after space
    printf("\n");
    for (int i = 10; i < space; i++) {
        printf(" ");
    }
    printf("%s (%s)\n", root->key, root->color == RED ? "R" : "B");

    // Process left child (below the parent)
    printTreeHelper(root->left, nil, space);
}

// Wrapper function to print the tree
void printTree(RedBlackTree *tree) {
    printf("\n--------------------");
    printTreeHelper(tree->root, tree->nil, 0);
    printf("--------------------\n");
}

// Test Cases
void testRedBlackTree() {
    RedBlackTree *tree = createRedBlackTree();

    // Test Case 0: Insert nodes
    insert(tree, createNode("Athena", "Wisdom"));
    insert(tree, createNode("Zeus", "Thunder"));
    insert(tree, createNode("Hades", "Underworld"));
    insert(tree, createNode("Poseidon", "Sea"));
    insert(tree, createNode("Hermes", "Messenger"));
    insert(tree, createNode("Demeter", "Harvest"));
    insert(tree, createNode("Dionysus", "Wine"));
    insert(tree, createNode("Ares", "War"));
    insert(tree, createNode("Artemis", "Hunt"));
    insert(tree, createNode("Hephaestus", "Forge"));
    printTree(tree);
    printf("\n");

    // free tree
    freeRedBlackTree(tree);

    // new tree
    tree = createRedBlackTree();

    // Test Case 1: Empty Tree
    printf("Test Case 1: Empty Tree\n");
    printf("Search for 'Athena': %s\n", search(tree, tree->root, "Athena") == tree->nil ? "Not Found" : "Found");
    deleteByKey(tree, "Athena"); // do nothing
    printf("\n");

    // Test Case 2: Single Node Tree
    printf("Test Case 2: Single Node Tree\n");
    insert(tree, createNode("Athena", "Wisdom"));
    printf("Inorder Walk:\n");
    inorderWalk(tree->root, tree->nil);
    printf("Search for 'Athena': %s\n", search(tree, tree->root, "Athena") == tree->nil ? "Not Found" : "Found");
    deleteByKey(tree, "Athena"); // delete the root
    printf("Inorder Walk after deletion:\n");
    inorderWalk(tree->root, tree->nil);
    printf("\n");

    // Test Case 3: Insertion of Duplicate Keys
    printf("Test Case 3: Insertion of Duplicate Keys\n");
    insert(tree, createNode("Athena", "Wisdom"));
    insert(tree, createNode("Athena", "New Value")); // update value or ignore
    printf("Inorder Walk:\n");
    inorderWalk(tree->root, tree->nil);
    printf("\n");

    // Test Case 4: Deletion of Non-Existent Keys
    printf("Test Case 4: Deletion of Non-Existent Keys\n");
    deleteByKey(tree, "NonExistentKey"); // do nothing
    printf("\n");

    // Test Case 5: Deletion of Root Node
    printf("Test Case 5: Deletion of Root Node\n");
    insert(tree, createNode("Athena", "Wisdom"));
    insert(tree, createNode("Zeus", "Thunder"));
    insert(tree, createNode("Hades", "Underworld"));
    printf("Inorder Walk before deletion:\n");
    inorderWalk(tree->root, tree->nil);
    printTree(tree);
    printf("\n");
    deleteByKey(tree, "Hades"); // delete the root
    printf("Inorder Walk after deletion:\n");
    inorderWalk(tree->root, tree->nil);
    printTree(tree);
    printf("\n");

    // Test Case 6: Deletion of Leaf Nodes
    printf("Test Case 6: Deletion of Leaf Nodes\n");
    insert(tree, createNode("Poseidon", "Sea"));
    printf("Inorder Walk before deletion:\n");
    inorderWalk(tree->root, tree->nil);
    printTree(tree);
    printf("\n");
    deleteByKey(tree, "Athena"); // delete the leaf node
    printf("Inorder Walk after deletion:\n");
    inorderWalk(tree->root, tree->nil);
    printTree(tree);
    printf("\n");

    // Test Case 7: Deletion of Nodes with One Child
    printf("Test Case 7: Deletion of Nodes with One Child\n");
    insert(tree, createNode("Hermes", "Messenger"));
    insert(tree, createNode("Demeter", "Harvest"));
    printf("Inorder Walk before deletion:\n");
    inorderWalk(tree->root, tree->nil);
    printTree(tree);
    deleteByKey(tree, "Hermes"); // delete node with one child
    printf("Inorder Walk after deletion:\n");
    inorderWalk(tree->root, tree->nil);
    printTree(tree);
    printf("\n");

    // Test Case 8: Deletion of Nodes with Two Children
    printf("Test Case 8: Deletion of Nodes with Two Children\n");
    insert(tree, createNode("Dionysus", "Wine"));
    insert(tree, createNode("Ares", "War"));
    printf("Inorder Walk before deletion:\n");
    inorderWalk(tree->root, tree->nil);
    printTree(tree);
    deleteByKey(tree, "Demeter"); // delete node with two children
    printf("Inorder Walk after deletion:\n");
    inorderWalk(tree->root, tree->nil);
    printTree(tree);
    printf("\n");

    // Test Case 9: Large Tree Stress Test
    printf("Test Case 9: Large Tree Stress Test\n");
    for (int i = 0; i < 100; i++) {
        char key[10], value[10];
        sprintf(key, "Key%d", i);
        sprintf(value, "Value%d", i);
        insert(tree, createNode(key, value));
    }
//  printTree(tree);
    printf("Inorder Walk after insertions:\n");
    inorderWalk(tree->root, tree->nil);
    for (int i = 0; i < 100; i++) {
        char key[10];
        sprintf(key, "Key%d", i);
        deleteByKey(tree, key);
    }
    printTree(tree);
    printf("Inorder Walk after deletions:\n");
    inorderWalk(tree->root, tree->nil);
    printTree(tree);
    printf("\n");
}

int main() {
    testRedBlackTree();
    return 0;
}
